File: scripts/prepare-hombres-fotos.py
import os
import sys
import glob
import shutil
import argparse
import cv2

# Rutas y constantes
DEFAULT_DIR = r"C:\Users\Evo-minidesk\Downloads\Hombres"
IMAGE_EXTS = ('.jpg', '.jpeg', '.png', '.webp', '.tiff', '.avif')

# Cargar el clasificador de rostros Haar Cascade
cascPath = cv2.data.haarcascades + "haarcascade_frontalface_default.xml"
faceCascade = cv2.CascadeClassifier(cascPath)

def get_face_ratio(img_path, verbose=False):
    try:
        # Usar cv2.imdecode para soportar rutas con tildes/espacios en Windows
        import numpy as np
        with open(img_path, 'rb') as f:
            chunk = f.read()
        chunk_arr = np.frombuffer(chunk, dtype=np.uint8)
        img = cv2.imdecode(chunk_arr, cv2.IMREAD_COLOR)

        if img is None:
            return 0.0

        img_area = img.shape[0] * img.shape[1]
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

        # Detectar rostros
        faces = faceCascade.detectMultiScale(
            gray,
            scaleFactor=1.1,
            minNeighbors=5,
            minSize=(50, 50)
        )

        if len(faces) == 0:
            return 0.0

        # Encontrar el rostro que ocupa mayor área
        max_face_area = 0
        for (x, y, w, h) in faces:
            area = w * h
            if area > max_face_area:
                max_face_area = area

        return max_face_area / img_area
    except Exception as e:
        if verbose:
            print(f"Error procesando {os.path.basename(img_path)}: {e}")
        return 0.0

def get_ficha_path(model_path):
    possible_names = ['Ficha', 'ficha', 'FICHA']
    for name in possible_names:
        full_path = os.path.join(model_path, name)
        if os.path.exists(full_path) and os.path.isdir(full_path):
            return full_path
    return None

def list_images(dir_path):
    if not os.path.exists(dir_path):
        return []
    files = [f for f in os.listdir(dir_path) if f.lower().endswith(IMAGE_EXTS) and os.path.isfile(os.path.join(dir_path, f))]
    return sorted(files)

def process_model(model_name, base_dir, dry_run=True, verbose=False):
    model_path = os.path.join(base_dir, model_name)
    ficha_path = get_ficha_path(model_path)

    if not ficha_path:
        print(f"  [SKIP] {model_name} - No tiene carpeta Ficha/ficha")
        return False

    portada_path = os.path.join(ficha_path, 'portada')
    
    # Check if portada exists and is not empty
    if os.path.exists(portada_path) and len(list_images(portada_path)) > 0:
        print(f"  [OK] {model_name} - Ya tiene portada pre-configurada.")
        return True

    ficha_files = list_images(ficha_path)

    if len(ficha_files) == 0:
        print(f"  [SKIP] {model_name} - Ficha vacía.")
        return False

    # Regla: <= 3 fotos, no crear portada
    if len(ficha_files) <= 3:
        print(f"  [INFO] {model_name} - Solo {len(ficha_files)} fotos. Omitiendo portada.")
        return True

    if verbose:
        print(f"  Analizando {len(ficha_files)} fotos en {model_name}/Ficha...")

    max_ratio = -1.0
    best_file = ficha_files[0]

    for f in ficha_files:
        full_path = os.path.join(ficha_path, f)
        ratio = get_face_ratio(full_path, verbose)
        
        if verbose:
            print(f"    - {f}: Ratio Rostro = {(ratio * 100):.2f}%")
            
        if ratio > max_ratio:
            max_ratio = ratio
            best_file = f

    source_path = os.path.join(ficha_path, best_file)
    target_path = os.path.join(portada_path, best_file)

    if not dry_run:
        if not os.path.exists(portada_path):
            os.makedirs(portada_path)
        shutil.move(source_path, target_path)

    print(f"  [DONE] {model_name} - Portada: {best_file} (Ratio: {(max_ratio * 100):.1f}%)")
    return True

def main():
    parser = argparse.ArgumentParser(description="Organiza las carpetas de hombres creando subcarpetas 'portada'")
    parser.add_argument("--dir", default=DEFAULT_DIR, help="Carpeta base (Hombres)")
    parser.add_argument("--models", help="Modelos específicos a procesar (por ejemplo: 'Adrian Estrada,Carlos Bonilla')")
    parser.add_argument("--dry-run", action="store_true", help="Simular sin mover archivos")
    parser.add_argument("--verbose", action="store_true", help="Imprimir progreso detallado")
    args = parser.parse_args()

    print("==========================================================")
    print("       Preparar Carpetas Hombres (OpenCV local)")
    print("==========================================================")
    print(f"  Modo:        {'DRY RUN (no mueve archivos)' if args.dry_run else 'LIVE (modificando carpetas)'}")
    print(f"  Directorio:  {args.dir}")
    print()

    if not os.path.exists(args.dir):
        print(f"Error: No existe el directorio {args.dir}")
        sys.exit(1)

    folders = [f for f in os.listdir(args.dir) if os.path.isdir(os.path.join(args.dir, f))]
    
    if args.models:
        filter_names = [x.strip().lower() for x in args.models.split(',')]
        folders = [f for f in folders if f.lower() in filter_names]
        print(f"  Filtro: Procesando solo {len(folders)} modelos\n")

    print(f"Encontrados {len(folders)} modelos para analizar...\n")

    count = 0
    for folder in folders:
        if process_model(folder, args.dir, args.dry_run, args.verbose):
            count += 1

    print("\n==========================================================")
    print(f"  Proceso terminado exitosamente. Analizados {count}/{len(folders)} modelos.")
    print("==========================================================\n")

if __name__ == "__main__":
    main()

File: scripts/prepare-hombres-fotos.mjs
/**
 * prepare-hombres-fotos.mjs
 * 
 * Este script corre LOCALMENTE. 
 * Su única función es organizar las carpetas de los modelos hombres para que 
 * tengan la misma estructura que las mujeres:
 * 
 * Antes:
 * {ModelName}/ficha/ 
 *   |-- foto1.jpg
 *   |-- foto2.jpg
 * 
 * Después:
 * {ModelName}/ficha/
 *   |-- portada/
 *   |     |-- mejor_foto_rostro.jpg  <-- (Auto-detectada)
 *   |-- foto1.jpg
 *   |-- foto2.jpg
 * 
 * Regla: Si hay 3 fotos o menos en "ficha", no crea portada (todas quedan en ficha).
 * 
 * Uso:
 * node scripts/prepare-hombres-fotos.mjs --dir="C:\Users\Evo-minidesk\Downloads\Hombres"
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import sharp from 'sharp';

// Face detection dependencies
import { Canvas, Image, ImageData } from '@napi-rs/canvas';
import * as faceapi from 'face-api.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const DRY_RUN = process.argv.includes('--dry-run');
const VERBOSE = process.argv.includes('--verbose');

function getArgValue(name) {
  const arg = process.argv.find((a) => a.startsWith(`--${name}=`));
  return arg ? arg.split('=')[1] ?? null : null;
}

const BASE_DIR = getArgValue('dir') ?? 'C:\\Users\\Evo-minidesk\\Downloads\\Hombres';
const FILTER_MODELS = getArgValue('models')?.split(',').map((s) => s.trim().toLowerCase()) ?? null;
const IMAGE_EXTENSIONS = /\.(jpg|jpeg|png|webp|tiff|avif)$/i;

// ======================================================================
// face-api.js setup
// ======================================================================
faceapi.env.monkeyPatch({ Canvas, Image, ImageData });

let modelsLoaded = false;
async function loadFaceModels() {
  if (modelsLoaded) return;
  const modelsPath = path.join(__dirname, 'models');
  if (VERBOSE) console.log(`Cargando modelos de IA desde ${modelsPath}...`);
  await faceapi.nets.tinyFaceDetector.loadFromDisk(modelsPath);
  modelsLoaded = true;
}

async function getFaceRatio(imagePath) {
  try {
    const processedBuffer = await sharp(imagePath)
      .rotate() 
      .jpeg() 
      .toBuffer();

    const img = new Image();
    img.src = processedBuffer;

    if (img.width === 0 || img.height === 0) return 0;

    const detections = await faceapi.detectAllFaces(img, new faceapi.TinyFaceDetectorOptions());
    if (!detections || detections.length === 0) return 0;

    let maxFaceArea = 0;
    for (const det of detections) {
      const faceArea = det.box.width * det.box.height;
      if (faceArea > maxFaceArea) {
        maxFaceArea = faceArea;
      }
    }

    const imgArea = img.width * img.height;
    return maxFaceArea / imgArea;
  } catch (err) {
    if (VERBOSE) console.warn(`Error al detectar rostro en ${path.basename(imagePath)}: ${err.message}`);
    return 0;
  }
}

// ======================================================================
// File System Helpers
// ======================================================================
function listImages(dirPath) {
  if (!fs.existsSync(dirPath)) return [];
  return fs
    .readdirSync(dirPath)
    .filter((f) => {
      const full = path.join(dirPath, f);
      return fs.statSync(full).isFile() && IMAGE_EXTENSIONS.test(f);
    })
    .sort();
}

function getFichaPath(modelPath) {
  const possibleNames = ['Ficha', 'ficha'];
  for (const name of possibleNames) {
    const fullPath = path.join(modelPath, name);
    if (fs.existsSync(fullPath)) return fullPath;
  }
  return null;
}

// ======================================================================
// Procesar un Modelo
// ======================================================================
async function processModel(folderName) {
  const modelPath = path.join(BASE_DIR, folderName);
  const fichaPath = getFichaPath(modelPath);

  if (!fichaPath) {
    console.log(`  [SKIP] ${folderName} - No tiene carpeta Ficha/ficha`);
    return false;
  }

  const portadaPath = path.join(fichaPath, 'portada');
  if (fs.existsSync(portadaPath) && listImages(portadaPath).length > 0) {
    console.log(`  [OK] ${folderName} - Ya tiene portada/ configurada.`);
    return true;
  }

  const fichaFiles = listImages(fichaPath);

  if (fichaFiles.length === 0) {
    console.log(`  [SKIP] ${folderName} - Ficha vacía.`);
    return false;
  }

  // Regla: <= 3 fotos, no crear portada
  if (fichaFiles.length <= 3) {
    console.log(`  [INFO] ${folderName} - Solo ${fichaFiles.length} fotos. Omitiendo portada.`);
    return true;
  }

  // Cargar modelos de IA solo cuando sean necesarios
  await loadFaceModels();

  let maxRatio = -1;
  let bestCoverIndex = 0;

  for (let i = 0; i < fichaFiles.length; i++) {
    const fullPath = path.join(fichaPath, fichaFiles[i]);
    const ratio = await getFaceRatio(fullPath);
    
    if (ratio > maxRatio) {
      maxRatio = ratio;
      bestCoverIndex = i;
    }
  }

  const coverFile = fichaFiles[bestCoverIndex];
  const sourcePath = path.join(fichaPath, coverFile);
  const targetPath = path.join(portadaPath, coverFile);

  if (!DRY_RUN) {
    if (!fs.existsSync(portadaPath)) {
      fs.mkdirSync(portadaPath, { recursive: true });
    }
    // Mover archivo
    fs.renameSync(sourcePath, targetPath);
  }

  console.log(`  [DONE] ${folderName} - Portada seleccionada: ${coverFile} (Ratio: ${(maxRatio * 100).toFixed(1)}%)`);
  return true;
}

// ======================================================================
// Main
// ======================================================================
async function main() {
  console.log('==========================================================');
  console.log('       Preparar Carpetas Hombres (Auto-Portada local)     ');
  console.log('==========================================================');
  console.log(`  Modo:        ${DRY_RUN ? 'DRY RUN (no mueve archivos)' : 'LIVE'}`);
  console.log(`  Directorio:  ${BASE_DIR}`);
  console.log('');

  if (!fs.existsSync(BASE_DIR)) {
    console.error(`Error: No existe el directorio: ${BASE_DIR}`);
    process.exit(1);
  }

  let folders = fs
    .readdirSync(BASE_DIR)
    .filter((f) => fs.statSync(path.join(BASE_DIR, f)).isDirectory());

  if (FILTER_MODELS) {
    folders = folders.filter((f) => FILTER_MODELS.includes(f.toLowerCase().trim()));
    console.log(`  Filtro: Procesando solo ${folders.length} modelos\n`);
  }

  console.log(`Encontrados ${folders.length} modelos para analizar...\n`);

  let count = 0;
  for (const folder of folders) {
    const success = await processModel(folder);
    if (success) count++;
  }

  console.log('\n==========================================================');
  console.log(`  Proceso terminado. Analizados ${count}/${folders.length} modelos.`);
  console.log('==========================================================\n');
}

main().catch(err => {
  console.error("Error fatal:", err);
  process.exit(1);
});
